File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    if (list1 === null && list2 === null) {
        return null
    }
    let list3 = {}
    
    let node1 = list1
    let node2 = list2
    let referenceNode = list3
    
    while (node1 !== null || node2 !== null) {
        if (node1 && node2) {
            if (node1.val <= node2.val) {
                 referenceNode.val = node1.val
                node1 = node1.next
            } else {
                 referenceNode.val = node2.val
                node2 = node2.next
            } 
        } else {
            if (node1) {
                referenceNode.val = node1.val
                node1 = node1.next
            }
            if (node2) {
                 referenceNode.val = node2.val
                node2 = node2.next
            }
        }
        if (!node1 && !node2) {
            referenceNode.next = null
        } else {
            referenceNode.next = {}
            referenceNode = referenceNode.next     
        }
       
    }
    referenceNode.next = null
    return list3
};

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        //if list1 is null
        if (list1 == null) { return list2; }
        if (list2 == null) { return list1; }
        //return list2
        //if list2 is null
        //return list1
        
      //compare list1val with list2val
        //if list1val is less than list2val
        if (list1.val < list2.val) {
            list1.next = mergeTwoLists(list1.next, list2);
            return list1;
        } else {
            list2.next = mergeTwoLists(list1, list2.next);
            return list2;
        }
        //set list1val's next property to equal the merge of list1val.next with list2val
        //return list1
        //if list2val is less than list1val
        //set list2val next property to equal the merge of list2val.next with list1val
        //return list2
    }
}